src/runner.rs

```diff
@@ -61,19 +68,107 @@ pub async fn run_with_config(
         .validate()?
         .build();
 
-    let (shutdown_tx, shutdown_rx) = watch::channel::<()>(());
+    let (shutdown_tx, shutdown_rx) = watch::channel(());
+    let (ready_tx, _ready_rx) = oneshot::channel();
     tokio::spawn(async move {
         // Don't unwrap in order to ensure that we execute
         // any subsequent shutdown tasks.
         signal::ctrl_c().await.ok();
         shutdown_tx.send(()).ok();
     });
 
-    if let Err(err) = server.run(shutdown_rx).await {
+    if let Err(err) = server.run(ready_tx, shutdown_rx).await {
         info!(log, "Shutting down with error"; "error" => %err);
         Err(Error::from(err))
     } else {
         info!(log, "Shutting down");
         Ok(())
     }
 }
+
+/// Start and run a proxy. Any passed in [`FilterFactory`]s are included
+/// alongside the default filter factories.
+pub async fn test(
+    base_log: slog::Logger,
+    mut testsuite: Testsuite,
+    filter_factories: impl IntoIterator<Item = DynFilterFactory>,
+) -> Result<(), Error> {
+    use once_cell::sync::Lazy;
+
+    static FEEDBACK_LOOP: Lazy<SocketAddr> = Lazy::new(|| {
+        let socket = std::net::UdpSocket::bind((Ipv4Addr::UNSPECIFIED, 0)).unwrap();
+        let local_addr = socket.local_addr().unwrap();
+        std::thread::spawn(move || loop {
+            let mut packet = [0; 0xffff];
+            let (_, addr) = socket.recv_from(&mut packet).unwrap();
+            let length = packet
+                .iter()
+                .position(|&x| x == 0)
+                .unwrap_or_else(|| packet.len());
+            let packet = &packet[..length];
+            eprintln!("Received: {:?}", packet);
+            socket.send_to(packet, addr).unwrap();
+        });
+
+        local_addr
+    });
+
+    let log = base_log.new(o!("source" => "run"));
+    let mut receiver = *FEEDBACK_LOOP;
+    receiver.set_ip(Ipv4Addr::LOCALHOST.into());
+
+    {
+        let endpoints = testsuite.config.source.get_static_endpoints_mut().unwrap();
```

Comment:
I'd prefer we didn't have any concept of mutability in our config, I think we can copy the testconfig and insert the exact `source` we want to use for test?
Coming to think of it, I think this means that all endpoint info in the config are ignored currently (until we have support to point the test to an endpoint)? if so we should call it out in the docs

Reply:
> Coming to think of it, I think this means that all endpoint info in the config are ignored currently (until we have support to point the test to an endpoint)? if so we should call it out in the docs

I can add docs about it. On using the real endpoints, that would be simple to add, the question is whether that would be useful. As working on your machine you're not always going to be able to access the same IPs that are in the config if it's not publicly accessible.

Reply:
can we copy the config rather than have the `get_static_endpoints_mut` function?

> can add docs about it. On using the real endpoints, that would be simple to add, the question is whether that would be useful. As working on your machine you're not always going to be able to access the same IPs that are in the config if it's not publicly accessible.

Yeah I think that would be useful otherwise we won't be able to do anything beyond unit tests. In the future we'd want to run test against an actual deployment e before rolling out changes for example where we know its IP